test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,668 @@
+    pthread_join(thread, NULL);
+
+    res = check_result(alloc_mem_recv.ptr, gold_data, size);
+
+    allocator_send->free(&alloc_mem_send);
+    allocator_recv->free(&alloc_mem_recv);
+
+    pop_and_release_primary_context(cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_user(const allocator_t *allocator_send,
+                                     const allocator_t *allocator_recv,
+                                     const void *gold_data, size_t size,
+                                     CUdevice cu_dev)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUcontext primary_ctx, user_ctx;
+    int res;
+
+    retain_and_push_primary_context(cu_dev);
+
+    alloc_mem_send = allocator_send->alloc(size);
+    alloc_mem_recv = allocator_recv->alloc(size);
+
+    cuda_memcpy((void*)alloc_mem_send.ptr, gold_data, size);
+
+    CUDA_CALL(cuCtxPopCurrent(&primary_ctx));
+    CUDA_CALL(cuCtxCreate(&user_ctx, 0, cu_dev));
+
+    mpi_pingpong(alloc_mem_send.ptr, alloc_mem_recv.ptr, size);
+
+    CUDA_CALL(cuCtxDestroy(user_ctx));
+    CUDA_CALL(cuCtxPushCurrent(primary_ctx));
+
+    res = check_result(alloc_mem_recv.ptr, gold_data, size);
+
+    allocator_send->free(&alloc_mem_send);
+    allocator_recv->free(&alloc_mem_recv);
+
+    pop_and_release_primary_context(cu_dev);
+
+    return res;
+}
+
+static int test_alloc_user_send_prim(const allocator_t *allocator_send,
+                                     const allocator_t *allocator_recv,
+                                     const void *gold_data, size_t size,
+                                     CUdevice cu_dev)
+{
+    CUcontext user_ctx;
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    int res;
+
+    CUDA_CALL(cuCtxCreate(&user_ctx, 0, cu_dev));
+
+    alloc_mem_send = allocator_send->alloc(size);
+    alloc_mem_recv = allocator_recv->alloc(size);
+
+    cuda_memcpy((void*)alloc_mem_send.ptr, gold_data, size);
+
+    retain_and_push_primary_context(cu_dev);
+    mpi_pingpong(alloc_mem_send.ptr, alloc_mem_recv.ptr, size);
+    pop_and_release_primary_context(cu_dev);
+
+    res = check_result(alloc_mem_recv.ptr, gold_data, size);
+
+    allocator_send->free(&alloc_mem_send);
+    allocator_recv->free(&alloc_mem_recv);
+
+    CUDA_CALL(cuCtxDestroy(user_ctx));
+
+    return res;
+}
+
+static int test_alloc_prim_send_other_prim(const allocator_t *allocator_send,
+                                           const allocator_t *allocator_recv,
+                                           const void *gold_data, size_t size,
+                                           CUdevice cu_dev)
+{
+    int res = 0, rank, dev_count, i;
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUdevice cu_dev_alloc, cu_dev_comm;
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    CUDA_CALL(cuDeviceGetCount(&dev_count));
+
+    for (i = 0; i < dev_count; ++i) {
+        CUDA_CALL(cuDeviceGet(&cu_dev_alloc, (i + rank) % dev_count));
```

Comment:
maybe iterate over all possible combinations instead?

Reply:
Left only one compbination, since the test allocates on one gpu and send using another. Iterating through all gpus doesn't improve coverage.